Added QR_Code data flip support, flip and retry after first ECC failure

genciberisha opened the pull request:
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/1046

Fixes  #23249, fixes  #23155, fixes #20724

Added quirc_flip() method to horizontally flip the data.
When the decoder fails to ECC the data we flip the image, and try one more time.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

genciberisha commented:
@alalek The test data for this PR is in : [opencv_extra/pull/1046](https://github.com/opencv/opencv_extra/pull/1046).

genciberisha commented:
@alalek @AleksandrPanov Is there something left to do in this PR from my side ?

AleksandrPanov commented:
@genciberisha I found this issue in the [quirc library](https://github.com/dlbeer/quirc): 
[QR code ECC failure #88](https://github.com/dlbeer/quirc/issues/88)
[support horizontally flipped QR-codes according to ISO 18004:2015 #89](https://github.com/dlbeer/quirc/pull/89)

Your solution to the problem is correct. Test images must be converted to jpg to take up less space.

asmorkalov commented:
The PR contains 2 tests that are dependent. It's not stable solution for testing.

genciberisha commented:
@AleksandrPanov Converting these images to **JPG** decreases the size a little but also with a lot off quality loss.
I managed to compress these **PNG** file to under **3kb** size, hope that counts.

---

@asmorkalov Agreed, I removed the generation test and replaced it with a simple validation test.
Please note because of the nature of the bug we need to validate the data at the end, so i saw it reasonable to

AleksandrPanov reviewed (approved):
@genciberisha, thank you
I consider this solution to be optimal, because it's like [quirc's](https://github.com/dlbeer/quirc/pull/89/files#:~:text=%7D-,void%20quirc_flip(struct%20quirc_code%20*code),-%7B) solution.

genciberisha commented:
@alalek Seems there isn't anything left here to do, thank you.

asmorkalov commented:
@AleksandrPanov Could you fix it?
